### PR TITLE
Video metadata fe

### PR DESCRIPTION
**Pull request by KarolDenst**

### CODE REVIEW

File: frontend/src/app/core/services/video.service.ts [hunk 1]
```diff
@@ -17,4 +18,10 @@ export class VideoService {
     
     return this.httpClient.get<VideoListDto>(`${this.videoPageWebAPIUrl}/user/videos`, { params: params });
   }
+
+  getVideoMetadata(id: string): Observable<VideoMetadataDto> {
+    let params = new HttpParams().set('id', id);
+    
+    return this.httpClient.get<VideoMetadataDto>(`${this.videoPageWebAPIUrl}/video-metadata`, { params: params });
```
Comment (Marvvvin5) [inline frontend/src/app/core/services/video.service.ts]: dont we need authorization?

File: backend/YouTubeV2.Application/Services/VideoServices/VideoService.cs [hunk 2]
```diff
@@ -48,5 +48,27 @@ public async Task SetVideoProcessingProgressAsync(Video video, ProcessingProgres
             video.ProcessingProgress = processingProgress;
             await _context.SaveChangesAsync(cancellationToken);
         }
+
+        public async Task<VideoMetadataDto> GetVideoMetadataAsync(Guid id, CancellationToken cancellationToken = default)
+        {
+            var video = await _context.Videos.Include(video => video.Tags).SingleAsync(video => video.Id == id, cancellationToken);
```
Comment (Marvvvin5) [inline backend/YouTubeV2.Application/Services/VideoServices/VideoService.cs]: what if video doesnt exist? Single throws exception then is it handled?

File: backend/YouTubeV2.Application/Services/VideoServices/VideoService.cs [hunk 3]
```diff
@@ -48,5 +48,27 @@ public async Task SetVideoProcessingProgressAsync(Video video, ProcessingProgres
...
+            var user = await _context.Users.SingleAsync(user => user.Videos.Contains(video), cancellationToken);
```
Comment (Marvvvin5) [inline backend/YouTubeV2.Application/Services/VideoServices/VideoService.cs]: cant we just include the user from video?

File: frontend/src/app/pages/video/video.component.html [hunk 4]
```diff
@@ -1,4 +1,57 @@
...
+            </div>
+          </div>
+          <div class="flex font-bold text-white border-round m-2">
+            <div class="p-col">
+              <p-button label="Subscribe"></p-button>
+            </div>
+          </div>   
+        </div>
+        <div class="videoDescription">
+          {{ videoMetadata.viewCount }} views {{ videoMetadata.uploadDate | timeAgo }} ago
```
Comment (KarolDenst) [inline frontend/src/app/pages/video/video.component.html]: timeAgo does not work will have to fix

File: backend/YouTubeV2.Application/Services/VideoServices/VideoService.cs [hunk 5]
```diff
@@ -48,5 +48,27 @@ public async Task SetVideoProcessingProgressAsync(Video video, ProcessingProgres
             video.ProcessingProgress = processingProgress;
             await _context.SaveChangesAsync(cancellationToken);
         }
+
+        public async Task<VideoMetadataDto> GetVideoMetadataAsync(Guid id, CancellationToken cancellationToken = default)
```
Comment (Marvvvin5) [inline backend/YouTubeV2.Application/Services/VideoServices/VideoService.cs:56]: i think getting the videometadata should update the viewcount. I have no other idea how to do it differently

File: backend/YouTubeV2.Application/Services/VideoServices/VideoService.cs [hunk 6]
```diff
@@ -48,5 +48,27 @@ public async Task SetVideoProcessingProgressAsync(Video video, ProcessingProgres
...
+                thumbnail.ToString(),
+                user.Id,
+                user.UserName,
+                video.ViewCount,
+                video.Tags.Select(tag => tag.ToString()).ToList(),
+                video.Visibility.ToString(),
+                video.ProcessingProgress.ToString(),
+                video.UploadDate.DateTime,
+                video.EditDate.DateTime,
+                video.Duration);
```
Comment (Marvvvin5) [inline backend/YouTubeV2.Application/Services/VideoServices/VideoService.cs:90]: where do we set the duration of the video? I think we just dont maybe it requires a fix

File: backend/YouTubeV2.Application/Services/VideoServices/VideoService.cs (hunk 5 above)
Comment (Marvvvin5) [inline backend/YouTubeV2.Application/Services/VideoServices/VideoService.cs:56]: probably we are missing the logic of not allowing other users to get info about private videos. I believe it is also missing in getting video stream so you can fix it

File: backend/YouTubeV2/Controllers/VideoController.cs [hunk 7]
```diff
@@ -82,6 +81,12 @@ public async Task<ActionResult> UploadVideoAsync(Guid id, [FromForm] IFormFile v
             return StatusCode(StatusCodes.Status202Accepted);
         }
 
+        [HttpGet("video-metadata")]
```
Comment (Marvvvin5) [inline backend/YouTubeV2/Controllers/VideoController.cs:86]: authorization?

File: backend/YouTubeV2.Api.Tests/VideoControllerTests/GetVideoMetadataAsyncTests.cs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+                .Returns(new ClaimsPrincipal(new ClaimsIdentity(new Claim[] { new(ClaimTypes.Role, Role.Simple) })));
+            _webApplicationFactory = Setup.GetWebApplicationFactory(_userService.Object);
+
+            var config = _webApplicationFactory.Services.GetService<IConfiguration>();
+            var connection = config!.GetConnectionString("Db");
+            await Setup.ResetDatabaseAsync(connection!);
+        }
+
+        [TestMethod]
+        public async Task TestMethod1Async()
```
Comment (Marvvvin5) [inline backend/YouTubeV2.Api.Tests/VideoControllerTests/GetVideoMetadataAsyncTests.cs]: TestMethod1Async

File: backend/YouTubeV2.Api.Tests/VideoControllerTests/GetVideoMetadataAsyncTests.cs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        {
+            // ARRANGE
+            var httpClient = _webApplicationFactory.CreateClient();
```
Comment (Marvvvin5) [inline backend/YouTubeV2.Api.Tests/VideoControllerTests/GetVideoMetadataAsyncTests.cs]: doesnt it need disposal

File: backend/YouTubeV2.Api.Tests/VideoControllerTests/GetVideoMetadataAsyncTests.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    [TestClass]
+    public class GetVideoMetadataAsyncTests
+    {
+        private WebApplicationFactory<Program> _webApplicationFactory = null!;
+        private readonly Mock<IUserService> _userService = new();
+
+        [TestInitialize]
+        public async Task Initialize()
+        {
+            _userService
```
Comment (Marvvvin5) [inline backend/YouTubeV2.Api.Tests/VideoControllerTests/GetVideoMetadataAsyncTests.cs:32]: do we need it?

Comment (KarolDenst) [reply backend/YouTubeV2.Api.Tests/VideoControllerTests/GetVideoMetadataAsyncTests.cs:32]: I think that now we need it after i added validation

**Comment by KarolDenst** [comment]
Did not test it yet

File: backend/YouTubeV2.Api.Tests/VideoControllerTests/GetVideoAsyncTests.cs [hunk 11]
```diff
@@ -38,7 +43,9 @@ public async Task GetVideoAsyncWithRangeHeaderShouldReturnPartOfTheFile()
             // ARRANGE
             int from = 2, to = 6;
             string testToken = "testToken";
-            using HttpClient httpClient = _webApplicationFactory.CreateClient();
+            using HttpClient httpClient = _webApplicationFactory
+                .WithAuthentication(ClaimsProvider.WithRoleAccessAndUserId(Role.Creator, Guid.NewGuid().ToString()))
```
Comment (Marvvvin5) [inline backend/YouTubeV2.Api.Tests/VideoControllerTests/GetVideoAsyncTests.cs:47]: if its new id anyway do we need it?

File: backend/YouTubeV2.Application/Exceptions/ForbiddenException.cs [hunk 12]
```diff
@@ -0,0 +1,24 @@
+﻿using System.Runtime.Serialization;
+
+namespace YouTubeV2.Application.Exceptions
+{
+    [Serializable]
```
Comment (Marvvvin5) [inline backend/YouTubeV2.Application/Exceptions/ForbiddenException.cs:5]: why is it serializable?

**Review by Marvvvin5 — APPROVED** [review]

File: backend/YouTubeV2.Application/Exceptions/ForbiddenException.cs (hunk 12 above)
Comment (KarolDenst) [reply backend/YouTubeV2.Application/Exceptions/ForbiddenException.cs:5]: it was auto generated

File: backend/YouTubeV2.Api.Tests/VideoControllerTests/GetVideoAsyncTests.cs (hunk 11 above)
Comment (KarolDenst) [reply backend/YouTubeV2.Api.Tests/VideoControllerTests/GetVideoAsyncTests.cs:47]: I dont remember why but yes 
image

**Pull request merged**